### DISCUSSION TITLE
Introduce remappability for terrain surface and terrain edge objects.

### DISCUSSION THREAD

**Discussion by Xtreme97-RCT** [OpenRCT2/OpenRCT2#16014] [Feature requests] [answered]
Now that the NSF is in develop I've begun fiddling with custom terrain objects and have been thinking that a useful feature would be to add remappability for custom terrain edges and surfaces. It can be introduced as a "hasPrimaryColour" option under the properties section of the object.json file, same as with scenery objects. The coloured grid surfaces are already being remapped from the same set of sprites within g1.dat with the only change being the colour specified.

This would require a change to the land tool to include an area for the remappable colour selection to go. I made a couple of quick mockups in paint of how this could be implemented:

![image](https://user-images.githubusercontent.com/64152940/143593461-ae4fa2fc-ba1f-49a1-820e-15c48e3e065f.png)

![image](https://user-images.githubusercontent.com/64152940/143593507-d421a203-5205-4e55-b761-504347338493.png)

**Comment by duncanspumpkin** [comment]
There are 5 bytes of free space in the surface element that could be used for this. Is there anything else though that might be wanted to become part of the surface element as once allocated we can't use it for anything else.

**Reply by karst** [reply]
I don't know if the OpenLoco larger edge sprites would use any of those bytes, but those would add so much more detailing. I think this would be a nice addition, but something would need to be made to convert the current to a new colourable elements, so all are one element and we don't have redundancies/duplicate elements. 
The surfaces and edges that should be recolourable would be the grids and checkerboards for surfaces and stucco's for edges.

**Comment by Broxzier** [comment]
Just to note, the grid is already recolourable through the code, but cannot currently be saved:
![Disco floor](https://cdn.discordapp.com/attachments/359815052930580481/702614463538528256/2018-03-25_01-19-45_-_openrct2_disco.gif)

**Comment by danielmeinert** [comment]
That's why we thought it might be not too hard to implement it. I wasn't aware that the memory space for surface elements was so tight though. With 32 colors selectable it would take 10 bits to store the information if there was surface and edge colorable, correct?

I personally would love to see it, as it opens a lot possibilities, without having to add a new object for any color variation of surface/edge you come up with.

**Comment by Gymnasiast** [comment, marked as answer]
https://github.com/OpenRCT2/OpenRCT2/pull/27151
